File: CSharpCodeBase/components/character/possibleactions.lua
local Class = require 'utils.hump.class'
local Component = require 'components.component'

local PossibleActions = Class({__includes=Component, init=function(self, entity)
  Component.init(self, entity)
  self.actions = {}
  self.defaultAction = 'attack'
  self:LoadActions()
  self.entity.possibleActions = self
end})

function PossibleActions:LoadActions()
  self.actions = self.entity.config:Get('action') or {}
end

function PossibleActions:GetAction(idx)
  if idx < 1 then
    return self.actions[1] or self.defaultAction
  elseif idx > #self.actions then
    return self.actions[#self.actions] or self.defaultAction
  else
    return self.actions[idx] or self.defaultAction
  end
end

function PossibleActions:SetActions(actions)
  self.actions = actions
end

return PossibleActions

File: CSharpCodeBase/components/character/possibleactions.cs
using System.Collections.Generic;
using UnityEngine;

namespace MainGame
{
    public class possibleactions : MonoBehaviour
    {
        private List<string> actions;
        private string defaultAction;

        public void Awake()
        {
            actions = new List<string>();
            defaultAction = "attack";
            LoadActions();
        }

        public void LoadActions()
        {
            actions = new List<string>(GetComponent<config>().Get("action") as object[]);
        }

        public string GetAction(int idx)
        {
            if (actions.Count == 0)
            {
                return defaultAction;
            }
            else if (idx < 0)
            {
                return actions[0];
            }
            else if (idx >= actions.Count)
            {
                return actions[actions.Count - 1];
            }
            else
            {
                return actions[idx];
            }
        }

        public void SetActions(string[] actions)
        {
            this.actions = new List<string>(actions);
        }
    }
}
